### DISCUSSION TITLE
pickling closures/lambdas fails in an ipyparallel context

### DISCUSSION THREAD

**Discussion by ftalbrecht** [pymor/pymor#1215] [Ideas]
Currently, pickling fails in ipyparallel when using the reduction error analysis with a custom error norm involving lambdas, closures or the like. This could be remedied by using [dill with ipyparallel](https://github.com/ipython/ipyparallel/blob/master/examples/Using%20Dill.ipynb), which seems to require execution of `use_dill` on each worker.

- Am I the first to stumble upon this problem, do we already have a fix for this, @sdrave?
- Do we want to support `dill` with `ipyparallel`?

**Comment by renefritze** [comment]
We already have [custom pickling ](https://docs.pymor.org/2020.1.1/generated/pymor.core.html?highlight=pickle#module-pymor.core.pickle).

```
In [1]: from pymor.core.pickle import *

In [2]: g = dumps(lambda : 1)

In [3]: h = loads(g)

In [4]: h
Out[4]: <function <lambda>()>

In [5]: h()
Out[5]: 1
```

**Reply by ftalbrecht** [reply]
True, I know about that one (though I never tested it in an ipyparallel environment). I am in particular interested if/how we want to setup the ipyparallel worker pool in pyMOR to allow advanced pickling (be it ours or the `dill`, I don't care about that).

**Reply by renefritze** [reply]
You can customize ipyparallel's serialization to use whatever you like IIRC. So you can try that with our pickling first.

**Reply by renefritze** [reply]
Did you work something out for your problem, @ftalbrecht ?

**Reply by sdrave** [reply]
FYI, I vaguely remember checking out `dill` before writing something myself. Back in the days, it did not work very well, but I can't remember what the problem was. It's been a while, so it might work better now.
